arch/rv32i/src/pmp.rs

```diff
@@ -486,210 +838,1388 @@ impl<const MAX_AVAILABLE_REGIONS_OVER_TWO: usize> kernel::platform::mpu::MPU
         permissions: mpu::Permissions,
         config: &mut Self::MpuConfig,
     ) -> Result<(), ()> {
-        let region_num = config.app_memory_region.unwrap_or(0);
-
-        let (region_start, _) = match config.regions[region_num] {
-            Some(region) => region.location(),
-            None => {
-                // Error: Process tried to update app memory MPU region before it was created.
-                return Err(());
-            }
-        };
+        let region_num = config.app_memory_region.get().ok_or(())?;
 
-        let app_memory_break = app_memory_break as usize;
+        let mut app_memory_break = app_memory_break as usize;
         let kernel_memory_break = kernel_memory_break as usize;
 
-        // Out of memory
+        // Ensure that the requested app_memory_break complies with PMP
+        // alignment constraints, namely that the region's end address is 4 byte
+        // aligned:
+        if app_memory_break % 4 != 0 {
+            app_memory_break += 4 - (app_memory_break % 4);
+        }
+
+        // Check if the app has run out of memory:
         if app_memory_break > kernel_memory_break {
             return Err(());
         }
 
-        // Get size of updated region
-        let region_size = app_memory_break - region_start as usize;
-
-        let region = PMPRegion::new(region_start, region_size, permissions);
-
-        config.regions[region_num] = Some(region);
+        // If we're not out of memory, update the region configuration
+        // accordingly:
+        config.regions[region_num].0 = permissions.into();
+        config.regions[region_num].2 = app_memory_break as *const u8;
         config.is_dirty.set(true);
 
         Ok(())
     }
 
     fn configure_mpu(&self, config: &Self::MpuConfig) {
-        // Is the PMP already configured for this app?
-        let last_configured_for_this_app = self
-            .last_configured_for
-            .map_or(false, |last_id| last_id == config.id);
-
-        // Skip PMP configuration if it is already configured for this app and the MPU
-        // configuration of this app has not changed.
-        if !last_configured_for_this_app || config.is_dirty.get() {
-            for (x, region) in config.regions.iter().enumerate() {
-                let disable_val = (csr::pmpconfig::pmpcfg::r0::CLEAR
-                    + csr::pmpconfig::pmpcfg::w0::CLEAR
-                    + csr::pmpconfig::pmpcfg::x0::CLEAR
-                    + csr::pmpconfig::pmpcfg::a0::OFF)
-                    .value;
-
-                let (region_shift, other_region_mask) = if x % 2 == 0 {
-                    (0, 0xFFFF_0000)
-                } else {
-                    (16, 0x0000_FFFF)
-                };
+        if !self.last_configured_for.contains(&config.id) || config.is_dirty.get() {
+            self.pmp.configure_pmp(&config.regions).unwrap()
+        }
```

Comment:
This code currently always reconfigures the PMP, it's missing the code to remember whether a given `MpuConfig` has already been configured:

```suggestion
        if !self.last_configured_for.contains(&config.id) || config.is_dirty.get() {
            self.pmp.configure_pmp(&config.regions).unwrap()
            config.is_dirty.set(false);
            self.last_configured_for.set(config.id);
        }
```